### ISSUE TITLE
rrssb 2.0 outline

### ISSUE THREAD

**Issue by dbox** [kni-labs/rrssb#113]
- New dev folder structure: `/assets`, `/dist`
- Ajax loading of inline svgs
- New plugin style JS call w/ options
- Total counts per #49

Obviously this is a major update and will have breaking changes. 1.x.x wil also still be available.

**Comment by connorwyatt** [comment]
By the way I've pushed my changes to my fork of rrssb and you can have a look if you like.

They dont resize yet because those functions arent in the file but I will add those in at some point

**Comment by connorwyatt** [comment]
https://github.com/connorwyatt/rrssb/tree/2.0

**Comment by connorwyatt** [comment]
There's no count yet either but that will be added

**Comment by AdamPS** [comment]
#103 ended with "I'll use the 2.0 thread from now on" so I'll pop a couple of questions / requests in here.  Context is that I use Drupal and have recently become a co-maintainer of the associated module https://www.drupal.org/project/rrssb.

General requests:
* Consider the "deliver mechanism".  Make it as easy as possible for downstream consumers to integrate changes.  Currently the index.html is a manual integration into our PHP file.  If you could deliver a PHP array that would be ideal (possibly you could use the PHP to make your index.html automatically).  Taking the SVG out of index.html will be a big help in itself.  The URLs in the PHP would ideally have placeholders such as [title] and [url].
* Allow extensibility by users of the library - in particular adding of new buttons.  This is easy in 1.0, so just need not to lose it when moving to an SVG file or similar.
* Be ready for success and a large number of buttons.  Basically take care of bloat from unused buttons.  In 1.0 the CSS for every button is included in the delivery and this will gradually become significant bloat.  It would be simpler if there were one CSS file without any button-specific directives.  Then the button-specific CSS would ideally be delivered alongside the SVG and URL information collectively as "button data" (PHP?) from which an index.html sample can be generated.  The Drupal module can take the button data in and generate appropriate files based on which buttons the user selects.
* Keep the library light-weight rather than getting bloated with many features - that's one of the key advantages that attracts people to it!

**Comment by AdamPS** [comment]
Specific question.  I don't understand how you can both "use AJAX/CDN to get the SVG" and "avoid bloat of downloading unwanted buttons".  Surely the client side script will fetch the full SVG from the CDN?

What I had expected was that the SVG would reside on the server, and the CMS (Drupal/Wordpress) could then see which buttons are configured and only pull in the appropriate parts of the SVG.

Simple hand-crafted sites would likely be happy with simply fetching the whole SVG so this is fine as a fallback behaviour.

**Comment by AdamPS** [comment]
Many thanks for all the hard work!

**Comment by connorwyatt** [comment]
To answer your final question first, each svg is downloaded separately and
only if the button is present so they aren't all downloaded, reducing the
bloat that would be required if all the svgs were included in the js script.

**Comment by connorwyatt** [comment]
To answer your previous questions, you could use php to insert the values
by using php variables in the script tag in your HTML (not in a separate js
file as php won't run on it).

You can add your own buttons too but you will need to make an svg, and then
get the API URL for the social network. Then you can code it into the js
like the others. The reason it isn't easier is because they all have
different URL structures.

Finally, all the CSS is quite small because most of the styles apply to all
buttons, only the colour of the button is different if my memory serves so
you can't really reduce the size.

It will still be lightweight, just easier to use.

**Comment by dbox** [comment]
"Consider the "deliver mechanism". Make it as easy as possible for downstream consumers to integrate changes. Currently the index.html is a manual integration into our PHP file. If you could deliver a PHP array that would be ideal (possibly you could use the PHP to make your index.html automatically). Taking the SVG out of index.html will be a big help in itself. The URLs in the PHP would ideally have placeholders such as [title] and [url]."

The new plugin js call will address this. PHP will not be used.

```
<body>

  <div class="rrssb-container></div>

  <script>

    var options = {
      socialNetworks: [ "facebook", "twitter", "linkedin", "email" ],
      showCount: true
    };

    $(".rrssb-container").rrssb( options );

  </script>

</body>

```

**Comment by AdamPS** [comment]
Well be careful with "each svg is downloaded separately" because I think it could impact the page load speed - each file to download adds an overhead: bytes sent to establish a new socket and also the socket transfers slowly when it starts up due to discovering the available speed.

Maybe better for the server to run the smarts and serve a single file containing the right icon set? Or at least allow this as an option?  I.e. a config parameter that could be set to point to URL of an SVG container or similar.

**Comment by AdamPS** [comment]
@dbox the plugin js call sounds good!

**Comment by AdamPS** [comment]
Thanks again for listening.  One more general consideration: what about the fallback if scripts aren't enabled?  I think it ought to be at least some sort of list of buttons that are links and contain text.  This is potentially important for:
* search engines and bots
* screen readers
* older/low-end devices
* bugs/compatibility problems meaning scripts don't load

I gather that kni-labs write high-end flashy websites, which is great.  But I'm more into generating simple small websites fairly frequently.  They don't have a lot of personal attention each so simple is good!  Some of my clients have 8-year old devices themselves!

**Comment by AdamPS** [comment]
It looks to me as if the the CSS is currently close to 500 bytes per button - nearly 50% of the SVG size.

Some other well-established social share solutions have around 100 buttons.

**Comment by connorwyatt** [comment]
I agree with @dbox, no php will be used but you can use php to insert the
social networks into the js options. That would be up to you though.

I'll look into the overhead of establishing multiple connections but
without multiple connections it would need php or another server side
script to do the job.

Remember that with the CSS the base styles are most of the file, the
specific styles are only a tiny part, meaning that adding new social
networks would add minimal code to the CSS.

**Comment by connorwyatt** [comment]
I don't personally work with kni labs but I do work with websites with a
lot of js, and you'll be glad to know that seo is not a problem as google
now runs js and indexes.

For older devices, a simpler solution might be your best bet (e.g. the idea
you had in another thread about moderately responsive buttons).

I personally don't write code for old browsers, I just use a piece of code
that let's older browser users know that their browser is out of date and
that the site may not work properly. I'm not a big fan of supporting ie8 or
below.

If you want to write a fallback for non js users then you could do that but
that isn't something I would personally do.

**Comment by connorwyatt** [comment]
By the way with regards to the fallback, the rrssb container gets emptied before adding the buttons (in the 2.0 I'm writing) so you could add a fallback in there and if the script runs it will remove the fallback

**Comment by Marc-pi** [comment]
@AdamPS 
you have great library to manage compat : modernizr, respond, html5shiv etc...
just use them. rrssb must work at state of art on stabilized techs.

yes, all serious search engines are reading now JS, google being at front of them

**Comment by AdamPS** [comment]
Thanks all for continuing to listen.  I guess I'm a bit old fashioned on scripts - I'll bow to your wisdom.

Regarding the CSS size though please can you double-check?  I just tested and my result was that 67% of the file was the specific parts.  Presumably this number will increase over time as more buttons are added?

**Comment by connorwyatt** [comment]
Apologies, I haven't looked yet, I will now

**Comment by dbox** [comment]
We may be able to rewrite the each loop to be a little more efficient.

**Comment by connorwyatt** [comment]
Just looked, I was thinking of the scss rather than the css. I'm not aware of a way round it though so it's going to have to stay like that. You could always remove rules you don't need

**Comment by connorwyatt** [comment]
@dbox I was just thinking that but I don't think you could unless you change some of the colours to be the same, eg Facebook and LinkedIn but I like the slightly different shades to be honest

**Comment by AdamPS** [comment]
It's possible that my aims are a little bit different from yours - but also I think I might be able to help you get the best of both worlds.  Please let me know if I'm not helping, and I'll leave you in peace!

**Comment by AdamPS** [comment]
CSS: close to 500 bytes actually contains 3 bits of real information: normal color; hover color; tiny hover color.  Why not have the js file store these three colours in a config array and then inject the CSS?

**Comment by AdamPS** [comment]
SVG files: I think there is a lot of evidence against fetching separate files: many articles; tools such as YSlow; feedback from users of the Drupal RRSSB module; the links I sent in my original issue - it seems like a step backwards from v1.0.

I expect it is faster to embed all the SVGs in the JS (like you say they are fairly compact).

But even better is to let the server-side code send only the data for the chosen buttons - I'll explain more in a mo...

**Comment by connorwyatt** [comment]
CSS injection isn't a bad idea, I'll test both methods of loading the SVG's and in the future I might look into writing a php version as you are right that server side includes would be a better bet, you'd still want the js though for the behaviour

**Comment by AdamPS** [comment]
HTML: so if I understand correctly, the base page will just be a placeholder which the JS populates with list items.  (Maybe I'm old-fashioned to prefer my HTML in-line then the script to style more like Superfish etc).  If you are going this way beware that users may currently be customising the HTML and so you might lose that flexibility.  

In particular don't assume that URL for sharing is the current page URL or title is current page title.  This breaks in a feed / if content is embedded / if the current URL isn't canonical etc.

Secondly, some of the sharing buttons require extra parameters.  E.g. youtube can't really share a page so the Drupal module allows configuration of the user's channel.  And pinterest needs an image to be supplied etc.

**Comment by connorwyatt** [comment]
Yes that is all being supplied, I've already written that in, check my 2.0 branch on my for of the project.

You can specify pinterest media, youtube link and github link

**Comment by connorwyatt** [comment]
And whilst I understand that some people prefer inline html, the jquery plugin call using .rrssb() means that you can set it up in seconds rather can copying html into your pages

**Comment by AdamPS** [comment]
Add you own buttons: you say it would be possible - so could you actually add a JS function for this?  I think it would pass: URL (with placeholders); either SVG file URL or SVG string; 3 colors as above.

This seems particularly neat, because then in the Drupal module, we can take the JS, delete all the config specific buttons (you could even ship a minimised version) and then server side PHP could add exactly the set of buttons that are needed using the above function (and passing SVG text rather than URL).

So potentially best of both worlds?  Use the larger script for a quick try out, it is entirely self-contained . Serious users that care about perf have the blazing fast lightweight version.  Anyone using a CMS gets that without extra effort because the plug-in or module does the work.

**Comment by AdamPS** [comment]
Definitely agree copying html into your pages is tedious - my last post hopefully covers that though.

**Comment by AdamPS** [comment]
"You can specify pinterest media, youtube link and github link" - nice.

Suggest you also need a function to supply URL and title, but it's optional the current page values are used by default.

**Comment by connorwyatt** [comment]
"Suggest you also need a function to supply URL and title, but it's
optional the current page values are used by default." - That is in there

**Comment by connorwyatt** [comment]
With regards to AJAX loading the SVGs I will test the speed of loading them all separately and putting them in an SVG sprite but then with an SVG sprite you will have unnecessary icons loading.

You could create your own sprite but that removes the ease of use of the plugin.

Secondly, the ajax loading of icons shouldn't be an issue because the page will still load and then the icons will "lazy load". You wont see a drop in performance, but it may take an extra 200ms to load the images. I dont think this is too much of a big deal. I'll test it though.

With regards to server processing, remember that this also stops caching and will slow the site down due to extra processing. Just a thought.

**Comment by connorwyatt** [comment]
Just looked at simultaneous connections, most modern browsers can have more than 12 concurrent requests, meaning that this shouldnt be an issue, plus AJAX is asychronous

**Comment by AdamPS** [comment]
"With regards to server processing, remember that this also stops caching and will slow the site down due to extra processing. Just a thought."

Well a way round that would be to get the server to generate an SVG file with a unique filename when the choice of buttons is confirmed and save to disk.  This could be cached by clients, would contain the minimum possible data, minimum possible number of files to transfer (each new file adds connection overhead possibly larger than the SVG, possible DNS look-up etc), and it all works just as well for custom buttons.  Could be a significant difference with a phone on a lousy mobile signal, but maybe most users won't notice as you say.

Anyway thanks very much for listening to my points.  Summing up, I think the two I'd suggest you to consider adopting are:
* CSS injection entirely client side.  Remove the per button CSS from the CSS file and instead store the colour values in the JS (roughly 5% the size) then inject.
* API function to add custom buttons.

I look forward to seeing the new version.

**Comment by connorwyatt** [comment]
I agree that that is an option but this plugin is designed to be JS only and with regards to CSS it is meant to be easy to manipulate too, so I'm not sure about it to be honest. If it is highly in demand to reduce the CSS size then I can implement it

**Comment by AdamPS** [comment]
Sure, no problem

**Comment by dbox** [comment]
Lets keep all styling to the SCSS for now. That's what it is for.

**Comment by connorwyatt** [comment]
+1

**Comment by connorwyatt** [comment]
http://connorwyatt.github.io/rrssb/ This is my current status with the 2.0 version, on GitHub pages. It doesn't resize yet but look at the speed of the icon load

**Comment by dbox** [comment]
Sweet, thanks. I know you're not ready for feedback yet but just looked
from phone and icons aren't loading.

Was nice and fast on desktop

**Comment by connorwyatt** [comment]
Okay no worries will look into it, it's better for me to know now :)

**Comment by connorwyatt** [comment]
If you have any more feedback let me know

**Comment by connorwyatt** [comment]
Just checked, theyre showing on mine, what phone are you using?

**Comment by dbox** [comment]
iPhone 6. Both chrome and safari

**Comment by connorwyatt** [comment]
I see, I cant check it as I don't have an iphone to hand, I use android. Can you debug on an iphone?

**Comment by connorwyatt** [comment]
All i can do is test it on a simulator and that isn't simulating the hardware and software, just the size http://mobiletest.me/iphone_5_emulator/#u=http://connorwyatt.github.io/rrssb/

**Comment by AdamPS** [comment]
Excellent, all sounds promising.  I'm going to unsubscribe from here.

We have a Drupal module for RRSSB here https://www.drupal.org/project/rrssb - maybe you could link it from your library page?  (I have recently become a co-maintainer but I wasn't the original author.)

I have posted an issue to upgrade to V2 of the library https://www.drupal.org/node/2493483.  If you remember, we'd appreciate a post there to let us know when the new version is ready.

Thanks!

**Comment by dbox** [comment]
@connorwyatt https://www.dropbox.com/s/kqf8p8y7n5eh92t/Screenshot%202015-05-22%2008.34.48.png?dl=0

@AdamPS We're already linking to it: https://github.com/kni-labs/rrssb/blob/master/README.md

**Comment by AdamPS** [comment]
D'oh sorry - thanks!

**Comment by connorwyatt** [comment]
Strange, I'll look into ajax/iphone issues and see what I can find

**Comment by connorwyatt** [comment]
Right. I have sorted the share count now, but as you will see, it is showing 1792 shares for the page because facebook thinks there have been 1792 shares...

http://graph.facebook.com/?id=http://connorwyatt.github.io/rrssb/

Any ideas how to fix this? I assume it will work on actual sites but I dont know what's going wrong

**Comment by connorwyatt** [comment]
Here is the link again if you need it http://connorwyatt.github.io/rrssb/

**Comment by dbox** [comment]
Same # as this: http://graph.facebook.com/?id=http://kurtnoble.com/labs/rrssb

**Comment by connorwyatt** [comment]
Strange... It is definitely using the connorwyatt.github url

**Comment by dbox** [comment]
Yeah, not sure why it's mirroring. Clearly the tool works: http://graph.facebook.com/?id=http://google.com

Which brings up a point about how we're going to handle the width of `.rrssb-count`. I think inline block is good, but we'll need it to push the other content over. I can help you with the styling of that whenever we pull your code in.

**Comment by connorwyatt** [comment]
It is working then, I’ll continue as it is

**Comment by connorwyatt** [comment]
The width is done by adding a style of calc(100% - 50px) to the rrssb icons container

**Comment by connorwyatt** [comment]
only if there is a count though

**Comment by dbox** [comment]
I think the facebook one is referencing the meta information instead of the actual url

**Comment by connorwyatt** [comment]
It can't be, open up your console and you will see that all the share counts are for the connorwyatt.github url

**Comment by dbox** [comment]
Yeah, I deleted the facebook url meta and uploaded to another server and it's still doing it:
http://dbox.us/rrssb2/

**Comment by dbox** [comment]
But doesn't do it locally

**Comment by connorwyatt** [comment]
Yeah my local one is showing 223

**Comment by connorwyatt** [comment]
It's only facebook too

**Comment by connorwyatt** [comment]
on the github pages one

**Comment by dbox** [comment]
Seems like there has to be something that's associating it specifically with kurtnoble.com/labs/rrssb

**Comment by connorwyatt** [comment]
Yeah definitely, I'll carry on going as I am and will have a look at some point

**Comment by dbox** [comment]
*The width is done by adding a style of calc(100% - 50px) to the rrssb icons container*

But if its a very long number it wont fit. Ideally the text should determine the width. But I can look into that part.

**Comment by connorwyatt** [comment]
That’s just for now, I’m planning to make it so that if it is long it will show 5K or 5M etc

**Comment by dbox** [comment]
ah. perfect.

**Comment by connorwyatt** [comment]
;)

**Comment by connorwyatt** [comment]
There you go, my github pages version has the K and M functionality now

**Comment by dbox** [comment]
@connorwyatt  Did you want me to pull into `2.0` or do you want to get a little further first?

**Comment by connorwyatt** [comment]
You can pull it in just let me check whether I need to push anything up.

I haven’t done anything with the sizing of the buttons yet so that is the next job that needs doing.

**Comment by bboerendans** [comment]
The thing why http://graph.facebook.com/?id=http://connorwyatt.github.io/rrssb/ and http://graph.facebook.com/?id=http://kurtnoble.com/labs/rrssb is giving the same count is that in both pages the meta property="og:url" have the same url :)

**Comment by dbox** [comment]
@bboerendans That's what we thought too but I put a page up with no og:url meta and it's still doing it:
http://dbox.us/rrssb2/

**Comment by connorwyatt** [comment]
Yeah we checked that @bboerendans and it’s not that, I have no idea what to do with it but I’ll look into it.

No URL is hard coded into the js so I can’t see why it’s playing up

**Comment by dbox** [comment]
I have an idea. Standby

**Comment by dbox** [comment]
@connorwyatt. Thanks! If you could bring in the current resizing that would be best.

**Comment by dbox** [comment]
@bboerendans @connorwyatt. Haha. So we were all right about the og-title. When I originally uploaded it, it had the og-title, then I took it off. Evidently all it takes is seeing it once and it caches it, even if you remove.

http://www.dbox.us/rrssb/ 

@connorwyatt - I think it's safe to assume script is all good. :)

**Comment by connorwyatt** [comment]
Cool, I would have been absolutely stumped otherwise.

Seems like a strangely permanent system though!

**Comment by dbox** [comment]
https://developers.facebook.com/tools/debug/og/object/

**Comment by dbox** [comment]
@connorwyatt You still working on stuff?

**Comment by connorwyatt** [comment]
Hi @dbox, at the minute I'm not as I've just started a new job, it's currently in a state where it loads the svgs using ajax and you can use it via jquery, it just doesn't resize.

Feel free to add in the resizing functions, the old ones should be fine, you would just need to change element names and classes. Otherwise I won't get round to it in the next month and it might be longer.

-------- Original message --------

**Comment by adamschwartz** [comment]
Hey @dbox, I really dig what y’all have built. I’m wondering if in 2.0 you might consider dropping jQuery as a dependency?

If you wanted to still support `$.fn.rrssb` you could bring that in if jQuery is detected. That way the advantages of using it with jQuery could be maintained while still allowing it to be used by websites which don’t use jQuery.

I’d be happy to help out if this is something you’re interested in doing.

**Comment by AdamPS** [comment]
So is your long term plan to drop v1.0?

I entered the debate further up the thread from the perspective of the co-maintainer of the Drupal integration.  I don't want to rerun the discussion again, but just a quick summary.  I can see that the drop-in zero integration option is attractive to many people.  But having the js generate the content is somewhat breaking the CMS model.  With v1.0, the HTML in Drupal starts off as an array and gradually gets expanded to the final HTML, with dozens of opportunities for customisation via hooks, templates, translations etc.  Some people are using these.  Also I believe our users are attracted by the fact the V1 module doesn't download a separate file per image - it is much more lightweight than comparable Drupal modules.  I know, you said V2.0 is super fast on a modern developer PC, but maybe it makes a difference over a slow broadband.  Certainly there are plenty of articles that say a single file is better performance.  I have a prototype roughly working with Drupal auto-generating a CSS file with SVGs as background images (hence better caching compared with inline SVG) although unfortunately can't do the micro format at the moment, so maybe I need to look at other ideas.

So net of all that is that I'm not sure how v2.0 would fit into the Drupal module.  Do you think there is any way that the two approaches can live side-by-side in the same project?  Clearly there is a fair overlap.  Or if we aren't sure about v2.0 would you suggest that we fork and go our own way?  I don't know what the original author of the Drupal module would say - maybe he would be more happy about the new V2.0 approach, but it's less good for my websites.

I can potentially help if we can find the side-by-side approach.  Although I mostly do php/CSS, and don't know much javascript, so I might not have the skills you are looking for.

**Comment by dbox** [comment]
@adamschwartz Personally, I don't mind dropping it as I don't do the JS updates. Would need to run by @joshuatuscan  and @connorwyatt. I guess I'd just need to see some data on what % of people use jquery and what the ramifications of removing it are.

**Comment by dbox** [comment]
@AdamPS 1.x will still be available, but I doubt we'll maintain it to the level that 2.0 will be. Definitely happy to have `rrssb-drupal` and `rrssb-wp` forks off the main repo, but I personally have zero interest in maintaining all the different versions. At a certain point, people like to build things differently, and that's okay!

**Comment by AdamPS** [comment]
@dbox Great, thanks it will help to have 1.0 around still.  I can understand it would be less well maintained.  Any updates to SVG files would perhaps be a really useful thing to transfer across and hopefully really easy to do as the files will stay in sync.

**Comment by adamschwartz** [comment]
@dbox I’m not sure why jQuery usage data would be necessary. Removing jQuery as a dependency would allow more people to use it, not fewer. That being said, the development time to remove it as a dependency (and continue to maintain it that way) is certainly real, but I’d be happy to help out there.

**Comment by dbox** [comment]
I just meant if 99% of people are using it, then probably not worth the
refactor, but I'm game.

We just gotta port those resize functions over first.

**Comment by adamschwartz** [comment]
Ahhh, I see, my mistake.

Yeah, as co-creator of [YMNNJQ](http://youmightnotneedjquery.com) I may be just a tad biased ;), but when a *library* can do everything it wants without jQuery, I typically advocate it. And like I said, I’d be happy to help out as I see it being very beneficial here. IMHO, a typical blog wouldn’t need jQuery since it’s just displaying static posts. Things like comments can be handled by Disqus, and Social Sharing can be handled by RRSSB :). One thing that drew me to RRSSB from the beginning was how lightweight it was as compared many other similar tools out there. At the time I didn’t realize that jQuery was a dep though. With jQuery added in, RRSSB’s size is an order of magnitude larger. Anywho, sorry again for my earlier confusion.

**Comment by AdamPS** [comment]
From the Drupal perspective, jQuery will very likely be loaded anyway.  But there is a certain amount of hassle to get v10 that is required here - the default is lower.

**Comment by joshuatuscan** [comment]
I think it’s a good idea to abandon jQuery as a dependency. Originally it was for speed and banking on the fact that most people will have jQuery running on their site already, however, vanilla JS would be better and I’m for it.

Like Daniel said, just need to port the functions away from using it.

**Comment by adamschwartz** [comment]
@AdamPS a great example of why dep management on the web is so hard. :+1: 

@joshuatuscan awesome! Let me know if I can help at all.

**Comment by dbox** [comment]
Checklist at the top of this thread has been updated with some tasks:) 

This one is the main priority right now:
> Resizing added back into new JS - @joshuatuscan / anyone else who can help

Not sure about @joshuatuscan's current availability, so if anyone else can dive in and help feel free!

**Comment by dbox** [comment]
Btw, 2.0 is DELIGHTFUL to work on thanks to all @connorwyatt's work on the gulp/browsersync setup.

**Comment by dbox** [comment]
Pinging to see availability on 2.0 work. @joshuatuscan you around at all? Anyone else want to dive in here? I'm currently slammed but would like to get 2.0 out the door if possible. I can help but can't take on the porting of resizing functionality.
